File: internal/limiter/sliding_window.go
package limiter

import (
	"context"
	"sync"
	"time"
)

// SlidingWindow implements Limiter using the sliding window log algorithm.
// It tracks the timestamp of each allowed request in a fixed-size ring buffer,
// evicting expired entries on each call to Allow.
type SlidingWindow struct {
	mu     sync.Mutex
	log    *RingBuffer
	limit  int64
	window time.Duration
}

// NewSlidingWindow returns a SlidingWindow ready for use.
// limit is the maximum number of requests allowed per window.
// window is the duration of the sliding time window.
func NewSlidingWindow(limit int64, window time.Duration) *SlidingWindow {
	newSlidingWindow := SlidingWindow{
		log:    NewRingBuffer(limit),
		limit:  limit,
		window: window,
	}
	return &newSlidingWindow
}

// Allow returns true if the request is within the rate limit, false if it should
// be throttled. It evicts expired entries before checking the limit.
// It is safe for concurrent use.
func (sw *SlidingWindow) Allow(ctx context.Context, key string) (bool, error) {
	sw.mu.Lock()
	defer sw.mu.Unlock()
	now := time.Now()
	cutoff := now.Add(-sw.window)

	if sw.log.Count() < int(sw.limit) {
		sw.log.Push(now)
		return true, nil
	}
	if sw.log.Oldest().Before(cutoff) {
		sw.log.EvictOldest()
		sw.log.Push(now)
		return true, nil
	}
	return false, nil
}

// RingBuffer is a fixed-capacity circular buffer of timestamps.
// It is used by SlidingWindow to track request times with no heap allocations
// after construction.
type RingBuffer struct {
	data  []time.Time
	head  int
	tail  int
	count int
}

// NewRingBuffer returns a RingBuffer with the given capacity.
func NewRingBuffer(capacity int64) *RingBuffer {
	newRingBuffer := RingBuffer{
		data:  make([]time.Time, capacity),
		head:  0,
		tail:  0,
		count: 0,
	}
	return &newRingBuffer
}

// Push writes t at the tail of the buffer and advances the tail.
func (rb *RingBuffer) Push(t time.Time) {
	rb.data[rb.tail] = t
	rb.tail = (rb.tail + 1) % len(rb.data)
	rb.count += 1
}

// Oldest returns the timestamp at the head of the buffer without removing it.
func (rb *RingBuffer) Oldest() time.Time {
	return rb.data[rb.head]
}

// EvictOldest advances the head of the buffer, discarding the oldest entry.
func (rb *RingBuffer) EvictOldest() {
	rb.head = (rb.head + 1) % len(rb.data)
	rb.count -= 1
}

// Count returns the number of live entries in the buffer.
func (rb *RingBuffer) Count() int {
	return rb.count
}

File: internal/store/lua/sliding_window.lua
-- Sliding window rate limiter.
-- Records the current request in a sorted set keyed by timestamp, evicts entries
-- outside the rolling window, and returns 1 if the request is within the limit,
-- 0 if it should be throttled. Time is sourced from Redis to avoid clock skew
-- across application nodes.
--
-- Returns {allowed, lua_us, remaining, reset_ms}: remaining is the quota
-- left in the rolling window after this request (never negative), reset_ms
-- the time until the oldest recorded request leaves the window — the moment
-- one unit of quota next comes back.
--
-- Key layout: this script reads and writes exactly KEYS[1] and never builds a
-- key name of its own. Redis Cluster only runs a script whose keys share one
-- hash slot, so one declared key makes it cluster-safe as-is. Do not add a
-- hash tag to the key prefix: keys are independent per rule and caller, and
-- a tag would pin all of them to one slot. Redis 7.4+: the script times
-- itself with os.clock(), because TIME is frozen for the whole execution.
-- See README "Single-key Lua scripts".
--
-- KEYS[1] - rate limit key (sorted set)
-- ARGV[1] - maximum number of requests allowed per window
-- ARGV[2] - window duration in milliseconds
-- ARGV[3] - optional; 'peek' reports the state a request arriving now would
--           see without recording one (the Admin API's quota inspection):
--           it counts the live window instead of evicting and adding.
local t0 = os.clock()

local time = redis.call('TIME')

local now_ms = time[1] * 1000 + math.floor(time[2]/1000)

local window_start = now_ms - tonumber(ARGV[2])

local limit = tonumber(ARGV[1])
local peek = ARGV[3] == 'peek'
local allowed = 0
local num_requests
local oldest

if peek then
    -- Entries at or before window_start are what the write path would evict;
    -- count strictly after it, and read the oldest survivor for reset.
    num_requests = redis.call('ZCOUNT', KEYS[1], '(' .. window_start, '+inf')
    if num_requests < limit then
        allowed = 1
    end
    oldest = redis.call('ZRANGEBYSCORE', KEYS[1], '(' .. window_start, '+inf', 'WITHSCORES', 'LIMIT', 0, 1)
else
    redis.call('ZREMRANGEBYSCORE', KEYS[1], '-inf', window_start)
    num_requests = redis.call('ZCARD', KEYS[1])
    if num_requests < limit then
        local member = tostring(time[1]) .. ':' .. tostring(time[2])
        redis.call('ZADD', KEYS[1], now_ms, member)
        redis.call('PEXPIRE', KEYS[1], tonumber(ARGV[2]))
        allowed = 1
        num_requests = num_requests + 1
    end
    oldest = redis.call('ZRANGE', KEYS[1], 0, 0, 'WITHSCORES')
end

local remaining = limit - num_requests
if remaining < 0 then
    remaining = 0
end

-- The oldest entry's score plus the window is when it will be evicted.
local reset_ms = 0
if oldest[2] then
    reset_ms = tonumber(oldest[2]) + tonumber(ARGV[2]) - now_ms
    if reset_ms < 0 then
        reset_ms = 0
    end
end

local lua_us = math.floor((os.clock() - t0) * 1000000)

return {allowed, lua_us, remaining, reset_ms}
